### PR TITLE
Dual-build for ESM and CJS

### PR DESCRIPTION
**Pull request by steveluscher**
# Summary

In this PR we introduce an ESM build. This should ensure compatibility with ESM build pipelines, including:

* Create React App
* Ember Embroider

…and all of the other ESM-based bundlers that folks have been having trouble with since v8.

This should be released as v9 because of how much change there is.

# Changes

* Everything is exported at the top level now, so that devs don't have to rely on the ability to dive into `dist/` to pick out one particular module (eg. `WebSocketBrowserImpl`). More about this, including an example, in `API.md`. Anything you don't use among the exports should be tree-shaken away by your optimizing compiler (eg. if you only use `WebSocket` then `Client` should get tree-shaken away).
* Each export is built into one file. This eliminates the class of problem where bundlers like Embroider had trouble importing dependencies in `dist/`.
* We bundle with Parcel now, and configure the entire build through `package.json`
* What gets published to npm now is just `dist/`, `README.md`, `LICENSE`, and `package.json`

# Test plan

## Unit tests

```
npm test
```

## `@solana/web3.js`

Applied [this diff](https://gist.github.com/steveluscher/c85ff3247e717b9d62e36dbd9bc486b8) and rebuilt. IIFE bundle worked.

## Create React App

1. Created a brand new CRA app
    ```
    npm create react-app test-app
    cd test-app
    npm add rpc-websockets@../rpc-websockets
    npm add @solana/web3.js@../solana-web3.js-git/packages/library-legacy
    ```
2. Added this code to `App.js`
    ```
    import { Connection } from "@solana/web3.js";
    import { Client } from "rpc-websockets";

    const ws = new Client("wss://echo.websocket.org");
    ws.on("open", function () {
        console.log("WebSocket Client Connected");
    });

    const c = new Connection("https://api.devnet.solana.com");
    c.getLatestBlockhash().then(console.log);
    c.onSlotChange(console.log);
    ```
    image

## Next.js with App Router

1. Created a brand new Next.js app
    ```
    npm create next-app test-app
    cd test-app
    npm add rpc-websockets@../rpc-websockets
    npm add @solana/web3.js@../solana-web3.js-git/packages/library-legacy
    ```
2. Added this code to `page.tsx`
    ```
    'use client';
    import { Connection } from "@solana/web3.js";
    import { Client } from "rpc-websockets";

    export default function Home() {
        useEffect(() => {
            const ws = new Client("wss://echo.websocket.org");
            ws.on("open", function () {
                console.log("WebSocket Client Connected");
            });

            const c = new Connection("https://api.devnet.solana.com");
            c.getLatestBlockhash().then(console.log);
            c.onSlotChange(console.log);
        }, []);
        /* ... */
    };
    ```
3. Ran `npm run dev`
    image

## Next.js with App Router

1. Created a brand new Next.js app with Pages router
    ```
    npm create next-app test-app
    cd test-app
    npm add rpc-websockets@../rpc-websockets
    npm add @solana/web3.js@../solana-web3.js-git/packages/library-legacy
    ```
2. Added this code to `_app.tsx`
    ```
    import { Connection } from "@solana/web3.js";
    import { Client } from "rpc-websockets";

    export default function App({ Component, pageProps }: AppProps) {
        useEffect(() => {
            const ws = new Client("wss://echo.websocket.org");
            ws.on("open", function () {
                console.log("WebSocket Client Connected");
            });

            const c = new Connection("https://api.devnet.solana.com");
            c.getLatestBlockhash().then(console.log);
            c.onSlotChange(console.log);
        }, []);
        /* ... */
    };
    ```
3. Ran `npm run dev`
    image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,10 +2,46 @@
...
+    "browser-bundle": {
+      "context": "browser",
+      "distDir": "dist/",
+      "outputFormat": "global",
+      "source": "src/index.browser-bundle.ts"
+    },
+    "types": {
+      "source": "src/index.ts"
+    }
+  },
```
Comment (steveluscher) [inline package.json]: This is the complete Parcel configuration. When you run `parcel build` it reads _this_ data to know how to behave.

File: package.json [hunk 2]
```diff
@@ -18,6 +54,9 @@
     "url": "git+https://github.com/elpheria/rpc-websockets.git"
   },
   "dependencies": {
+    "@types/uuid": "^8.3.4",
+    "@types/ws": "^8.2.2",
+    "buffer": "^6.0.3",
```
Comment (steveluscher) [inline package.json:47]: Downstream clients may actually need these dependencies if they're building for the browser, so I moved them out of `devDependencies`.

File: tsconfig.json [hunk 3]
```diff
@@ -1,7 +1,8 @@
 {
     "compilerOptions": {
         "outDir": "./dist",
-        "declaration": true,
+        "noEmit": true,
+        "declaration": false,
```
Comment (steveluscher) [inline tsconfig.json]: Only used for typechecking now, not for building.

**Comment by steveluscher** [comment]
Converting to draft for a moment; gotta figure out a way to build a typedef for the browser bundle (ie. it needs to omit `Server` from the exports)

**Comment by steveluscher** [comment]
OK, Parcel is not appropriate for this multi-package build (browser, node, iife). I've replaced it with the build pipeline that we use ourselves to build multiple packages, retested it, and added Next.js test plans for both App router and Pages router.

Ready to review.

File: tsup.config.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+const BROWSER_OPTIONS: Options = {
+    esbuildPlugins: [
+        polyfillNode({
+            globals: {
+                buffer: true,
+            },
+            polyfills: {
+                buffer: true,
+            },
```
Comment (steveluscher) [inline tsup.config.ts:28]: When building `browser-bundle.js`, include the `buffer` polyfill.

**Comment by mkozjak** [comment]
Amazing work, @steveluscher! Thanks much.
Tests are passing. I will squash-merge this one now.

**Review by mkozjak — APPROVED** [review]

**Comment by steveluscher** [comment]
I hope this fixes everyone’s build issues once and for all. Sorry for the wild ride, everyone. Once this works it should let y’all:

* Import just what you need for smaller bundles
* Build for the exact target you need for smaller bundles
* Work in ESM projects

**Pull request merged**